File: module-sdk/host.ts
/// Capability-gated host calls for Tier 2 iframe modules.
///
/// Each function wraps the postMessage protocol so module authors
/// never see raw envelopes. Calls return a Promise that rejects
/// with `HostError` on capability denial or unreachable host.
///
/// On the wire side every call goes shell → modulesd → policy check
/// → reply. Capability denial returns a typed `HostError` here so
/// modules can fall back gracefully.

import type {
  HostCall,
  HostReply,
  ModuleToHost,
  HostToModule,
  ErrorCode,
} from "./postmsg";

export class HostError extends Error {
  code: ErrorCode;
  constructor(code: ErrorCode, message: string) {
    super(message);
    this.code = code;
    this.name = "HostError";
  }
}

let nextRequestId = 1;
const pending = new Map<
  string,
  { resolve: (r: HostReply) => void; reject: (e: HostError) => void }
>();

/// Initialise the postMessage bridge. Must be called once at module
/// startup before any host call. Returns a promise that resolves on
/// the first `init` message from the shell.
export function bridgeInit(): Promise<{
  capabilities: import("./postmsg").Capabilities;
  theme: import("./postmsg").ThemeTokens;
}> {
  return new Promise((resolve) => {
    const handler = (ev: MessageEvent) => {
      const data = ev.data as HostToModule;
      if (data && data.type === "init") {
        window.removeEventListener("message", handler);
        resolve({ capabilities: data.capabilities, theme: data.theme });
      }
    };
    window.addEventListener("message", handler);
    sendUntyped({ type: "ready" });
  });
}

function sendUntyped(msg: ModuleToHost) {
  window.parent.postMessage(msg, "*");
}

function nextId(): string {
  return `req-${nextRequestId++}`;
}

function call(call: HostCall): Promise<HostReply> {
  return new Promise((resolve, reject) => {
    const requestId = nextId();
    pending.set(requestId, { resolve, reject });
    sendUntyped({ type: "host.call", requestId, call });
  });
}

// One global listener that demultiplexes host replies. Set up at
// module load; idempotent if `bridgeInit` is called multiple times.
window.addEventListener("message", (ev) => {
  const data = ev.data as HostToModule;
  if (!data || data.type !== "host.reply") return;
  const handler = pending.get(data.requestId);
  if (!handler) return;
  pending.delete(data.requestId);
  if (data.reply.type === "error") {
    handler.reject(new HostError(data.reply.code, data.reply.message));
  } else {
    handler.resolve(data.reply);
  }
});

export const graph = {
  /// Read Cypher query against the Knowledge Graph.
  /// Capability: `graph.read` allowlist on the queried namespace.
  async query(cypher: string): Promise<unknown[]> {
    const reply = await call({ type: "graph.query", cypher });
    if (reply.type !== "graph.result") {
      throw new HostError("internal", `unexpected reply: ${reply.type}`);
    }
    return JSON.parse(reply.rows);
  },

  /// Write Cypher.
  /// Capability: `graph.write` allowlist on the touched namespace.
  async write(cypher: string): Promise<unknown[]> {
    const reply = await call({ type: "graph.write", cypher });
    if (reply.type !== "graph.result") {
      throw new HostError("internal", `unexpected reply: ${reply.type}`);
    }
    return JSON.parse(reply.rows);
  },
};

export const network = {
  /// HTTP GET. URL host must be in the manifest's `network.allow`.
  async fetch(
    url: string,
    headers: Array<[string, string]> = [],
  ): Promise<{ status: number; body: Uint8Array }> {
    const reply = await call({ type: "network.fetch", url, headers });
    if (reply.type !== "network.body") {
      throw new HostError("internal", `unexpected reply: ${reply.type}`);
    }
    const binStr = atob(reply.bodyB64);
    const bytes = new Uint8Array(binStr.length);
    for (let i = 0; i < binStr.length; i++) bytes[i] = binStr.charCodeAt(i);
    return { status: reply.status, body: bytes };
  },
};

export const events = {
  /// Emit an event to the system Event Bus.
  /// Capability: `event_bus.publish` prefix allowlist.
  async emit(eventType: string, payload: Uint8Array): Promise<void> {
    const payloadB64 = btoa(String.fromCharCode(...payload));
    const reply = await call({ type: "events.emit", eventType, payloadB64 });
    if (reply.type !== "acked") {
      throw new HostError("internal", `unexpected reply: ${reply.type}`);
    }
  },
};

File: module-sdk/src/host.rs
/// Host imports surfaced to module authors.
///
/// On the WASM side these functions are linked at component-load time
/// against the `lunaris:host/*` interfaces defined in `wit/host.wit`;
/// on the native side (running tests in a host crate, no WASM) they
/// route through a thread-local mock that lets tests pre-program
/// answers. The signatures match between both paths so the same
/// module source compiles natively for unit tests and to WASM for
/// production.
///
/// The capability gating happens in `lunaris-modulesd`. Every host
/// import returns a typed `HostError` on denial; modules are expected
/// to handle that gracefully.

use thiserror::Error;

#[derive(Debug, Clone, Error, PartialEq, Eq)]
pub enum HostError {
    #[error("permission denied: {0}")]
    Denied(String),
    #[error("not found: {0}")]
    NotFound(String),
    #[error("network: {0}")]
    Network(String),
    #[error("internal: {0}")]
    Internal(String),
}

pub mod graph {
    use super::HostError;

    /// Read-only Cypher query against the Knowledge Graph. The host
    /// extracts the namespace from the query and matches it against
    /// the module's `graph.read` allowlist; mismatched queries return
    /// `HostError::Denied`.
    ///
    /// The shape of `rows` is JSON. Modules deserialise it with
    /// `serde_json` per their query.
    pub fn query(_cypher: &str) -> Result<String, HostError> {
        #[cfg(target_arch = "wasm32")]
        {
            // WASM path: link against lunaris:host/graph#query. The
            // wit-bindgen-generated stub goes here once S5 wires it
            // up alongside the Currency-Konverter dogfood. Until
            // then this is a deliberate compile-time placeholder
            // that fails closed.
            Err(HostError::Internal(
                "module-sdk: host/graph not linked yet (S5)".into(),
            ))
        }
        #[cfg(not(target_arch = "wasm32"))]
        {
            super::native_mock::graph_query(_cypher)
        }
    }

    /// Write to the Knowledge Graph. Same gating rules as `query`.
    pub fn write(_cypher: &str) -> Result<String, HostError> {
        #[cfg(target_arch = "wasm32")]
        {
            Err(HostError::Internal(
                "module-sdk: host/graph not linked yet (S5)".into(),
            ))
        }
        #[cfg(not(target_arch = "wasm32"))]
        {
            super::native_mock::graph_write(_cypher)
        }
    }
}

pub mod network {
    use super::HostError;

    /// HTTP GET. Body is returned raw. Capability: `network.allow`.
    pub fn fetch(_url: &str) -> Result<Vec<u8>, HostError> {
        #[cfg(target_arch = "wasm32")]
        {
            Err(HostError::Internal(
                "module-sdk: host/network not linked yet (S5)".into(),
            ))
        }
        #[cfg(not(target_arch = "wasm32"))]
        {
            super::native_mock::network_fetch(_url)
        }
    }
}

pub mod events {
    use super::HostError;

    /// Emit an event to the Lunaris Event Bus. Capability:
    /// `event_bus.publish` allowlist on the event type prefix.
    pub fn emit(_event_type: &str, _payload: &[u8]) -> Result<(), HostError> {
        #[cfg(target_arch = "wasm32")]
        {
            Err(HostError::Internal(
                "module-sdk: host/events not linked yet (S5)".into(),
            ))
        }
        #[cfg(not(target_arch = "wasm32"))]
        {
            super::native_mock::events_emit(_event_type, _payload)
        }
    }
}

pub mod log {
    /// Emit an info-level log line. Always allowed.
    pub fn info(message: &str) {
        #[cfg(target_arch = "wasm32")]
        {
            // wit-bindgen stub goes here in S5.
            let _ = message;
        }
        #[cfg(not(target_arch = "wasm32"))]
        {
            super::native_mock::log_info(message);
        }
    }
}

/// Native test mock. Lets tests pre-program responses for host
/// imports so module logic can be unit-tested without spinning up a
/// daemon. WASM builds skip this whole module.
#[cfg(not(target_arch = "wasm32"))]
pub mod native_mock {
    use super::HostError;
    use std::cell::RefCell;
    use std::collections::HashMap;

    thread_local! {
        static GRAPH_RESPONSES: RefCell<HashMap<String, Result<String, HostError>>> =
            RefCell::new(HashMap::new());
        static NETWORK_RESPONSES: RefCell<HashMap<String, Result<Vec<u8>, HostError>>> =
            RefCell::new(HashMap::new());
        static LOG_LINES: RefCell<Vec<String>> = RefCell::new(Vec::new());
    }

    pub fn set_graph_response(query: &str, response: Result<String, HostError>) {
        GRAPH_RESPONSES.with(|m| {
            m.borrow_mut().insert(query.to_string(), response);
        });
    }

    pub fn set_network_response(url: &str, response: Result<Vec<u8>, HostError>) {
        NETWORK_RESPONSES.with(|m| {
            m.borrow_mut().insert(url.to_string(), response);
        });
    }

    pub(super) fn graph_query(cypher: &str) -> Result<String, HostError> {
        GRAPH_RESPONSES.with(|m| {
            m.borrow()
                .get(cypher)
                .cloned()
                .unwrap_or_else(|| Err(HostError::NotFound(cypher.to_string())))
        })
    }

    pub(super) fn graph_write(cypher: &str) -> Result<String, HostError> {
        GRAPH_RESPONSES.with(|m| {
            m.borrow()
                .get(cypher)
                .cloned()
                .unwrap_or_else(|| Err(HostError::NotFound(cypher.to_string())))
        })
    }

    pub(super) fn network_fetch(url: &str) -> Result<Vec<u8>, HostError> {
        NETWORK_RESPONSES.with(|m| {
            m.borrow()
                .get(url)
                .cloned()
                .unwrap_or_else(|| Err(HostError::NotFound(url.to_string())))
        })
    }

    pub(super) fn events_emit(_event_type: &str, _payload: &[u8]) -> Result<(), HostError> {
        Ok(())
    }

    pub(super) fn log_info(message: &str) {
        LOG_LINES.with(|m| m.borrow_mut().push(message.to_string()));
    }

    /// Inspect logged lines. Useful for assertions in tests.
    pub fn captured_log_lines() -> Vec<String> {
        LOG_LINES.with(|m| m.borrow().clone())
    }

    /// Reset all mock state. Call at the start of each test so
    /// thread-local state from previous tests does not leak in.
    pub fn reset() {
        GRAPH_RESPONSES.with(|m| m.borrow_mut().clear());
        NETWORK_RESPONSES.with(|m| m.borrow_mut().clear());
        LOG_LINES.with(|m| m.borrow_mut().clear());
    }
}

#[cfg(not(target_arch = "wasm32"))]
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn graph_query_returns_mocked_response() {
        native_mock::reset();
        native_mock::set_graph_response(
            "MATCH (n) RETURN n",
            Ok("[{\"n\": 1}]".into()),
        );
        let rows = graph::query("MATCH (n) RETURN n").unwrap();
        assert_eq!(rows, "[{\"n\": 1}]");
    }

    #[test]
    fn graph_query_unmocked_returns_not_found() {
        native_mock::reset();
        let err = graph::query("MATCH (x) RETURN x").unwrap_err();
        assert!(matches!(err, HostError::NotFound(_)));
    }

    #[test]
    fn network_fetch_returns_mocked_body() {
        native_mock::reset();
        native_mock::set_network_response(
            "https://api.example.com/x",
            Ok(b"hello".to_vec()),
        );
        let body = network::fetch("https://api.example.com/x").unwrap();
        assert_eq!(body, b"hello");
    }

    #[test]
    fn network_fetch_unmocked_returns_not_found() {
        native_mock::reset();
        let err = network::fetch("https://nope.example.com").unwrap_err();
        assert!(matches!(err, HostError::NotFound(_)));
    }

    #[test]
    fn log_captures_lines() {
        native_mock::reset();
        log::info("module: hello");
        log::info("module: world");
        let lines = native_mock::captured_log_lines();
        assert_eq!(lines, vec!["module: hello", "module: world"]);
    }

    #[test]
    fn events_emit_is_silent_success() {
        native_mock::reset();
        events::emit("test.event", b"{}").unwrap();
    }
}
